drivers/display/display_framebuf.c

```diff
@@ -0,0 +1,251 @@
+ * should encounter no trouble. Avoid manipulating image data as byte arrays.
+ *
+ * Note that this implementation isn't particularly fast; it's aimed at
+ * simplicity. There's a lot of room for performance improvement.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <display.h>
+#include <display/framebuf.h>
+#include <string.h>
+
+#define GREY 0x00C0C0C0
+#define BLACK 0x00000000
+
+static int framebuf_blanking_on(const struct device *dev)
+{
+	return -ENOTSUP;
+}
+
+static int framebuf_blanking_off(const struct device *dev)
+{
+	return -ENOTSUP;
+}
+
+static void *framebuf_get_framebuffer(const struct device *dev)
+{
+	return NULL;
+}
+
+static int framebuf_set_brightness(const struct device *dev,
+				   const u8_t brightness)
+{
+	return -ENOTSUP;
+}
+
+static int framebuf_set_contrast(const struct device *dev,
+				 const u8_t contrast)
+{
+	return -ENOTSUP;
+}
+
+static int framebuf_set_pixel_format(const struct device *dev,
+				     const enum display_pixel_format format)
+{
+	struct framebuf_dev_data *data = FRAMEBUF_DATA(dev);
+
+	switch (format) {
+	case PIXEL_FORMAT_ARGB_8888:
+	case PIXEL_FORMAT_MONO01:
+	case PIXEL_FORMAT_MONO10:
+		data->format = format;
+		return 0;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static int framebuf_set_orientation(const struct device *dev,
+				    const enum display_orientation orientation)
+{
+	switch (orientation) {
+	case DISPLAY_ORIENTATION_NORMAL:
+		return 0;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static void framebuf_get_capabilities(const struct device *dev,
+				      struct display_capabilities *caps)
+{
+	struct framebuf_dev_data *data;
+
+	data = FRAMEBUF_DATA(dev);
+	caps->x_resolution = data->width;
+	caps->y_resolution = data->height;
+	caps->supported_pixel_formats = PIXEL_FORMAT_ARGB_8888 |
+					PIXEL_FORMAT_MONO01 |
+					PIXEL_FORMAT_MONO10;
+	caps->screen_info = 0;
+	caps->current_pixel_format = data->format;
+	caps->current_orientation = DISPLAY_ORIENTATION_NORMAL;
+}
+
+/*
+ * simple bit manipulation functions for display buffer read/write.
+ *
+ * we can't use the system-defined ones because they work on words,
+ * not bytes, which would introduce a dependency on endianness.
+ */
+
+static inline int test_bit(const void *buf, int bit)
+{
+	const u8_t *buf8 = buf;
+
+	return buf8[bit >> 3] & (1 << (7 - (bit & 7)));
+}
+
+static inline void set_bit(void *buf, int bit)
+{
+	u8_t *buf8 = buf;
+
+	buf8[bit >> 3] |= (1 << (7 - (bit & 7)));
+}
+
+static inline void reset_bit(void *buf, int bit)
+{
+	u8_t *buf8 = buf;
+
+	buf8[bit >> 3] &= ~(1 << (7 - (bit & 7)));
+}
+
+static int framebuf_write(const struct device *dev, const u16_t x,
+			  const u16_t y,
+			  const struct display_buffer_descriptor *desc,
+			  const void *buf)
+{
+	struct framebuf_dev_data *data = FRAMEBUF_DATA(dev);
+	u32_t *dst = data->buffer;
+	const u32_t *src;
+	int row, col, bit;
+	u32_t set_color, reset_color;
+
+	dst += x;
+	dst += (y * data->pitch);
+
+	switch (data->format) {
```

Comment:
Not super important, as our "read/write one pixel per function pointer indirection" API sucks to begin with, but for performance reasons this switch() ought to be at the initialization level, selecting format-specific callbacks for whatever got detected.

(Also, someone needs to sit down and write a proper blit API for devices to implement)

Reply:
I'm not sure I understand.  This switch() can't be done at initialization time, because the format of the user's display buffers (both for read and write) can be changed dynamically with a call to set_pixel_format().  Also, I'm not sure what you're saying about "one pixel at a time", the display read and write functions work with pixel buffers of arbitrary dimensions.  I'm lost.

Reply:
@andyross this specific API was intended to do partial buffer updates over slow busses (e.g. SPI/I2C). There is a direct API to get the actual framebuffer and perform the operations directly on the buffer which has not been implemented.